design/batching.md

```diff
@@ -0,0 +1,944 @@
+A _lot_ of things are gonna change, so we really want A) to avoid crazy large PRs that are a pain to review and B) to be able detect regressions (both correctness and performance) early on so they don't turn into long & painful investigations.
+
+Nothing set in stone obviously, but the following steps seem like a good start (roughly 1 step == 1 PR).  
+This entire section can pretty much be used verbatim as a tracking issue.
+
+1. Implement all the needed tests & benchmarks  
+We need to be able to check for regressions at every step, so make sure we have all the tests and benchmarks we need for that.  
+We should already be 95% of the way there at this point.
+
+1. Move `DataStore` sanity checks and formatting tools to separate files  
+`store.rs` is supposed to be the place where one can get an overview of all the datastructures involved in the store, except it has slowly become a mess over time and is now pretty much unreadable.
+
+1. Replace `MsgBundle` & `ComponentBundle` with the new types (`DataCell`, `DataRow`, `DataTable`, `EventId`, `BatchId`...)  
+No actual batching features nor any kind of behavior changes of any sort: just define the new types and use them everywhere.
+
+1. Pass entity path as a column rather than as metadata  
+Replace the current entity_path that is passed in the metadata map with an actual column instead. This will also requires us to make `EntityPath` a proper arrow datatype (..datatype, not component!!).
+
+1. Implement explicit number of instances  
+Introduce a new column for `num_instances`, integrate it in the store index and expose it in the store APIs.
+
+1. Fix splats all around (rs sdk, py sdk, re_query...)  
+Update the SDKs and `re_query` to properly make use of the new explicit `num_instances`.
+
+1. Get rid of component buckets altogether  
+Update the store implementation to remove component tables, remove the `get` APIs, introduce slicing on the write path, etc. Still no batching in sight!
+
+1. Implement the coalescing/accumulation logic in the SDK  
+Add the required logic/thread/timers/whatever-else in the SDKs to accumulate data and just send it all as many `LogMsg`s (i.e. no batching yet).
+
+1. Implement full-on batching  
+End-to-end: transport, storage, the whele shebang.
+
+1. Sort the batch before sending (`(event_id, entity_path)`)  
+Keep that in its own PR to keep track of the benchmarks.
+
+1. Implement new GC  
+The complete implementation; should close all existing GC issues.
+
+1. Dump directly from the store into an rrd file  
+No rebatching yet, just dump every event in its own `LogMsg`.
+
+1. Remove `LogMsg`s from `LogDb`  
+We shouldn't need to keep track of events outside the store past this point: clean it all up.  
+Reminder: the timeline widget keeps track of timepoints directly, not events.
+
+1. Rebatch aggressively while dumping to disk  
+
+1. Use arrow extension types to carry around component names  
+
+- Turn all of the above into a tracking issue
+- Get to work :>
```

Comment:
This plan is "replace one piece of the machine at the time while it is running". Did you consider the alternative: build a new machine on the side, and then swap it in when it is ready?

Reply:
Yes I did; and while I'm generally a major proponent of the nuclear approach, in this case we already have everything in place for a smooth incremental transition at very little extra cost, but with all the benefits of taking small steps at a time.